src/Mixin.php

```diff
@@ -0,0 +1,1961 @@
+<?php
+
+/**
+ * provides type inference and auto-completion for magic static methods of Assert.
+ */
+
+namespace Webmozart\Assert;
+
+use InvalidArgumentException;
+
+interface Mixin
+{
+    /**
+     * @psalm-pure
+     * @psalm-assert null|string $value
+     *
+     * @param mixed  $value
+     * @param string $message
+     *
+     * @throws InvalidArgumentException
+     */
+    public static function nullOrString($value, $message = '');
+
+    /**
+     * @psalm-pure
+     * @psalm-assert iterable<string> $value
```

Comment:
An important test to be written here: given an `ArrayIterator<mixed>`, the returned type (after asserting) should be `ArrayIterator<string>`, not `iterable<string>` (too generic)

Reply:
Can we actually do that? An arbitrary iterable can be passed here.

Reply:
yes, but *after* the assertion has gone through, is the original type retained or gone?

Reply:
That would be great, I'm not sure if you can type it like that in psalm though.

Reply:
Actually it looks it's good by default: https://psalm.dev/r/7687908084

Reply:
Its probably a good idea to add a test for it here, so we know if we accidentally broke an annotation somewhere (or if psalm itself broke something)

Reply:
@BackEndTea I will start covering everything with tests as long as we agree on everything else in this discussion.

I will mark this conversation as resolved (because everything will be indeed covered with tests before final merge).
Then it's one little change requested by Marco.

And one more conversation I need input from you @BackEndTea (ctrl+f by `Resolve conversation` and check all that are left).

Reply:
TIL: `@psalm-trace`